src/java/org/apache/cassandra/db/partitions/TriePartitionUpdate.java

```diff
@@ -0,0 +1,670 @@
+    }
+
+    /**
+     * Modify this update to set every timestamp for live data to {@code newTimestamp} and
+     * every deletion timestamp to {@code newTimestamp - 1}.
+     *
+     * There is no reason to use that except on the Paxos code path, where we need to ensure that
+     * anything inserted uses the ballot timestamp (to respect the order of updates decided by
+     * the Paxos algorithm). We use {@code newTimestamp - 1} for deletions because tombstones
+     * always win on timestamp equality and we don't want to delete our own insertions
+     * (typically, when we overwrite a collection, we first set a complex deletion to delete the
+     * previous collection before adding new elements. If we were to set that complex deletion
+     * to the same timestamp that the new elements, it would delete those elements). And since
+     * tombstones always wins on timestamp equality, using -1 guarantees our deletion will still
+     * delete anything from a previous update.
+     */
+    public TriePartitionUpdate withUpdatedTimestamps(long newTimestamp)
+    {
+
+        InMemoryTrie<Object> t = InMemoryTrie.shortLived();
+        try
+        {
+            t.apply(trie, new InMemoryTrie.UpsertTransformer<Object, Object>()
+            {
+                public Object apply(Object shouldBeNull, Object o)
+                {
+                    assert shouldBeNull == null;
+                    if (o instanceof RowData)
+                        return applyRowData((RowData) o);
+                    else
+                        return applyDeletion((DeletionInfo) o);
+                }
+
+                public RowData applyRowData(RowData update)
+                {
+                    LivenessInfo newInfo = update.livenessInfo.isEmpty()
+                                           ? update.livenessInfo
+                                           : update.livenessInfo.withUpdatedTimestamp(newTimestamp);
+                    // If the deletion is shadowable and the row has a timestamp, we'll force the deletion timestamp to be less
+                    // than the row one, so we should get rid of said deletion.
+                    Row.Deletion newDeletion = update.deletion.isLive() || (update.deletion.isShadowable() && !update.livenessInfo.isEmpty())
+                                               ? Row.Deletion.LIVE
+                                               : new Row.Deletion(new DeletionTime(newTimestamp - 1, update.deletion.time().localDeletionTime()),
+                                                                  update.deletion.isShadowable());
+
+                    Row updated = toRow(update, Clustering.EMPTY).updateAllTimestamp(newTimestamp);
+                    return new RowData(((BTreeRow)updated).getBTree(), newInfo, newDeletion);
+                }
+
+                public DeletionInfo applyDeletion(DeletionInfo update)
+                {
+                    MutableDeletionInfo mdi = update.mutableCopy();
```

Comment:
Nit: maybe check `update.isLive()` to avoid a new object creation in this relatively common case?

Reply:
Added.